File: typescript/src/core/hooks.ts
/**
 * Hooks for the realtime SDK: in-process lifecycle callbacks fired by the
 * session at four seams (``SessionStart``, ``PreToolUse``, ``PostToolUse``,
 * ``SessionEnd``), plus declarative server hooks the SERVER executes.
 *
 * Two kinds of hooks, one idea (when X happens, do Y), one ``hooks: [...]``
 * list on the agent:
 *
 * - **client hooks** — your callbacks, run in-process. Declare with the seam
 *   factories: ``sessionStart(fn)``, ``preToolUse(fn, {matcher})``,
 *   ``postToolUse(fn)``, ``sessionEnd(fn)``. List order is fold order.
 * - **server hooks** — wire config the server executes, working even if this
 *   process dies mid-call: a ``SilenceTimeout`` with a ``say`` or
 *   ``end_call`` action.
 *
 * Every seam hangs off something this process does, which is what makes it a
 * seam rather than a notification: ``SessionStart`` rewrites the config before
 * it is sent, ``PreToolUse`` gates a local handler, ``PostToolUse`` reads that
 * handler's outcome, and ``SessionEnd`` fires even when the socket dropped and
 * no frame ever arrived. Anything the SERVER did reaches you as an event —
 * ``on('user_speech_timeout')`` for a fired silence hook — never as a hook.
 *
 * Observer-grade by default; the two client-controlled seams honor
 * overrides — ``SessionStart`` may inject ``additionalContext`` into the
 * instructions and ``PreToolUse`` may deny or rewrite a local client-tool
 * call. A throwing hook is isolated and never breaks the session. Python
 * (``cosmo_ai.hooks``) is the reference for shape and semantics; the
 * matcher grammar and fold rules are pinned by the shared conformance
 * vectors shared across the SDKs.
 */

import { log } from './logger';
import type {
  SilenceTimeout,
  UserSpeechTimeoutEvent,
} from '../wire/types.gen';

import type { DisconnectReason } from './state';

/** A server-executed hook — wire config, not a callback. The family's
 *  extension point: becomes a union when a second server-hook kind lands. */
export type ServerHook = SilenceTimeout;

// A hook runs in-process on the session's hot path (SessionStart blocks
// session establishment; PreToolUse/PostToolUse are awaited inline in the
// tool-call RPC reply during a live voice turn). Nothing bounds a hook's
// runtime, so a slow hook (e.g. a network call) stalls that path; this only
// warns, it never cancels or times out a hook.
const SLOW_HOOK_WARN_THRESHOLD_MS = 200;

// ── Tool outcome (what PostToolUse observes) ───────────────────────────

export type ToolOutcome =
  | { kind: 'ok'; result: Record<string, unknown> | null }
  | { kind: 'error'; message: string }
  | { kind: 'denied'; reason: string };

// ── Per-event contexts ─────────────────────────────────────────────────

export type SessionStartContext = {
  // The session id does not exist until the handshake completes; read the
  // ``ready`` event for the started id.
  event: 'SessionStart';
};

export type PreToolUseContext = {
  event: 'PreToolUse';
  toolName: string;
  /** Read-only view; rewrite via ``PreToolUseResult.updatedArguments``. */
  arguments: Readonly<Record<string, unknown>>;
  sessionId: string;
};

export type PostToolUseContext = {
  event: 'PostToolUse';
  toolName: string;
  arguments: Record<string, unknown>;
  outcome: ToolOutcome;
  sessionId: string;
};

export type SessionEndContext = {
  event: 'SessionEnd';
  reason: DisconnectReason;
  detail: string | null;
  sessionId: string | null;
};

/** The server action a fired silence timeout performed (``say`` or
 *  ``end_call``), as reported on the wire frame. */
export type ServerHookAction = UserSpeechTimeoutEvent['action'];

// ── Per-event results ──────────────────────────────────────────────────

export type SessionStartResult = {
  additionalContext?: string | null;
};

export type PreToolUseResult = {
  permission?: 'allow' | 'deny';
  reason?: string;
  updatedArguments?: Record<string, unknown>;
};

/** Folded result of all PreToolUse hooks for one tool call. */
export type PreToolUseOutcome = {
  denied: boolean;
  reason: string | null;
  arguments: Record<string, unknown>;
};

export type SessionStartHook = (
  ctx: SessionStartContext,
) =>
  | SessionStartResult
  | null
  | undefined
  | void
  | Promise<SessionStartResult | null | undefined | void>;

export type PreToolUseHook = (
  ctx: PreToolUseContext,
) => PreToolUseResult | null | undefined | void | Promise<PreToolUseResult | null | undefined | void>;

export type PostToolUseHook = (ctx: PostToolUseContext) => void | Promise<void>;

export type SessionEndHook = (ctx: SessionEndContext) => void | Promise<void>;

// ── Matcher grammar ────────────────────────────────────────────────────

/** Normative matcher grammar, shared with the Python and Swift SDKs via
 *  ``hook-matcher-vectors.json``: glob-style
 *  ``*`` ``?`` ``[seq]`` ``[!seq]``, case-sensitive, matched against the
 *  full tool name, no path semantics. A malformed pattern is rejected at
 *  registration (``validateMatcher``) rather than reaching this function,
 *  so every pattern seen here is well-formed. */
export function toolNameMatches(toolName: string, pattern: string): boolean {
  return translateMatcher(pattern).test(toolName);
}

/** Reject an unterminated ``[...]`` group at hook-registration time.
 *
 *  The glob grammar treats a stray ``[`` as a literal character, so e.g.
 *  ``matcher="[delete_*"`` would silently never match any real tool name
 *  instead of erroring. For a ``PreToolUse`` deny matcher that is a silent
 *  fail-open (the guard never fires), so this fails loud instead. */
export function validateMatcher(pattern: string): void {
  let i = 0;
  const n = pattern.length;
  while (i < n) {
    if (pattern[i] === '[') {
      let j = i + 1;
      if (j < n && pattern[j] === '!') j += 1;
      if (j < n && pattern[j] === ']') j += 1;
      while (j < n && pattern[j] !== ']') j += 1;
      if (j >= n) {
        throw new Error(
          `malformed hook matcher ${JSON.stringify(pattern)}: unterminated '[' at index ${String(i)}`,
        );
      }
      i = j + 1;
    } else {
      i += 1;
    }
  }
}

const REGEXP_SPECIALS = /[.*+?^${}()|[\]\\]/g;

function translateMatcher(pattern: string): RegExp {
  let out = '';
  let i = 0;
  const n = pattern.length;
  while (i < n) {
    const c = pattern[i];
    i += 1;
    if (c === '*') {
      out += '.*';
    } else if (c === '?') {
      out += '.';
    } else if (c === '[') {
      let j = i;
      if (j < n && pattern[j] === '!') j += 1;
      if (j < n && pattern[j] === ']') j += 1;
      while (j < n && pattern[j] !== ']') j += 1;
      if (j >= n) {
        out += '\\[';
      } else {
        let stuff = pattern.slice(i, j).replace(/\\/g, '\\\\').replace(/\]/g, '\\]');
        if (stuff.startsWith('!')) stuff = `^${stuff.slice(1)}`;
        else if (stuff.startsWith('^')) stuff = `\\${stuff}`;
        out += `[${stuff}]`;
        i = j + 1;
      }
    } else {
      out += c.replace(REGEXP_SPECIALS, '\\$&');
    }
  }
  // ``s`` so ``*`` / ``?`` match newlines, mirroring Python's ``(?s:...)``.
  return new RegExp(`^(?:${out})$`, 's');
}

// ── Declared hooks + the seam factories ────────────────────────────────

export type HookEventName =
  | 'SessionStart'
  | 'PreToolUse'
  | 'PostToolUse'
  | 'SessionEnd';

type AnyHookCallback =
  | SessionStartHook
  | PreToolUseHook
  | PostToolUseHook
  | SessionEndHook;

/** One declared client hook: the seam it fires at, the callback, and (for
 *  the tool seams) an optional matcher. Built by the seam factories; list
 *  order in ``hooks: [...]`` is fold order. */
export class Hook {
  readonly event: HookEventName;
  /** @internal */
  readonly callback: AnyHookCallback;
  readonly matcher: string | null;

  /** @internal — construct via the seam factories. */
  constructor(event: HookEventName, callback: AnyHookCallback, matcher: string | null) {
    this.event = event;
    this.callback = callback;
    this.matcher = matcher;
    Object.freeze(this);
  }
}

/** Declare a ``SessionStart`` hook — may return a ``SessionStartResult``
 *  to inject ``additionalContext`` into the instructions. */
export function sessionStart(fn: SessionStartHook): Hook {
  return new Hook('SessionStart', fn, null);
}

/** Declare a ``PreToolUse`` hook — may deny or rewrite a local client-tool
 *  call. ``matcher`` restricts it to matching tool names (glob grammar); a
 *  malformed matcher throws here, not at session start. */
export function preToolUse(fn: PreToolUseHook, opts?: { matcher?: string }): Hook {
  if (opts?.matcher !== undefined) validateMatcher(opts.matcher);
  return new Hook('PreToolUse', fn, opts?.matcher ?? null);
}

/** Declare a ``PostToolUse`` observer, fired with the final ``ToolOutcome``
 *  of each local client-tool call. */
export function postToolUse(fn: PostToolUseHook, opts?: { matcher?: string }): Hook {
  if (opts?.matcher !== undefined) validateMatcher(opts.matcher);
  return new Hook('PostToolUse', fn, opts?.matcher ?? null);
}

/** Declare a ``SessionEnd`` observer, fired exactly once at teardown. */
export function sessionEnd(fn: SessionEndHook): Hook {
  return new Hook('SessionEnd', fn, null);
}

/** Split one unified ``hooks: [...]`` list into the client hooks (folded
 *  in-process, in list order) and the server hooks (wire config). Rejects
 *  anything that is neither. @internal */
export function resolveHooks(
  hooks: readonly (Hook | ServerHook)[] | undefined,
): { clientHooks: Hook[]; serverHooks: ServerHook[] } {
  const clientHooks: Hook[] = [];
  const serverHooks: ServerHook[] = [];
  for (const hook of hooks ?? []) {
    if (hook instanceof Hook) {
      clientHooks.push(hook);
    } else if (
      hook !== null &&
      typeof hook === 'object' &&
      // ``trigger`` is optional on the wire type; the required
      // ``timeout_seconds`` is the reliable discriminator.
      typeof (hook as ServerHook).timeout_seconds === 'number'
    ) {
      serverHooks.push(hook);
    } else {
      throw new Error(
        'hooks elements must be seam-factory Hooks or server hooks (SilenceTimeout)',
      );
    }
  }
  return { clientHooks, serverHooks };
}

// ── HookEngine ─────────────────────────────────────────────────────────

/** Dispatch engine over one agent's declared client hooks. Immutable —
 *  built from the resolved ``Hook`` list at session start; fold semantics
 *  are pinned by the shared hook-engine vectors. @internal */
export class HookEngine {
  private readonly sessionStartHooks: SessionStartHook[] = [];
  private readonly preToolUseHooks: { matcher: string | null; hook: PreToolUseHook }[] =
    [];
  private readonly postToolUseHooks: { matcher: string | null; hook: PostToolUseHook }[] =
    [];
  private readonly sessionEndHooks: SessionEndHook[] = [];

  constructor(hooks: readonly Hook[]) {
    for (const hook of hooks) {
      switch (hook.event) {
        case 'SessionStart':
          this.sessionStartHooks.push(hook.callback as SessionStartHook);
          break;
        case 'PreToolUse':
          this.preToolUseHooks.push({
            matcher: hook.matcher,
            hook: hook.callback as PreToolUseHook,
          });
          break;
        case 'PostToolUse':
          this.postToolUseHooks.push({
            matcher: hook.matcher,
            hook: hook.callback as PostToolUseHook,
          });
          break;
        case 'SessionEnd':
          this.sessionEndHooks.push(hook.callback as SessionEndHook);
          break;
      }
    }
  }

  /** Fold every SessionStart hook's ``additionalContext`` in list order;
   *  ``null`` when no hook contributed context. */
  async runSessionStart(ctx: SessionStartContext): Promise<string | null> {
    const chunks: string[] = [];
    for (const hook of this.sessionStartHooks) {
      const out = await callHook(hook, ctx);
      if (out === null || out === undefined) continue;
      if (out.additionalContext) chunks.push(out.additionalContext);
    }
    return chunks.length > 0 ? chunks.join('\n\n') : null;
  }

  async runPreToolUse(opts: {
    toolName: string;
    arguments: Record<string, unknown>;
    sessionId: string;
  }): Promise<PreToolUseOutcome> {
    let current: Record<string, unknown> = { ...opts.arguments };
    for (const { matcher, hook } of this.preToolUseHooks) {
      if (matcher !== null && !toolNameMatches(opts.toolName, matcher)) continue;
      const out = await callHook(hook, {
        event: 'PreToolUse',
        toolName: opts.toolName,
        arguments: Object.freeze({ ...current }),
        sessionId: opts.sessionId,
      });
      if (out === null || out === undefined) continue;
      if (out.permission === 'deny') {
        log.info('[realtime] hook denied tool', {
          tool: opts.toolName,
          reason: out.reason,
        });
        return {
          denied: true,
          // ``||`` not ``??``: an empty-string reason folds to the default,
          // matching Python's ``reason or "denied by hook"``.
          reason: out.reason || 'denied by hook',
          arguments: current,
        };
      }
      if (out.updatedArguments !== undefined) {
        current = { ...out.updatedArguments };
      }
    }
    return { denied: false, reason: null, arguments: current };
  }

  async runPostToolUse(ctx: PostToolUseContext): Promise<void> {
    for (const { matcher, hook } of this.postToolUseHooks) {
      if (matcher !== null && !toolNameMatches(ctx.toolName, matcher)) continue;
      await callHook(hook, ctx);
    }
  }

  async runSessionEnd(ctx: SessionEndContext): Promise<void> {
    for (const hook of this.sessionEndHooks) {
      await callHook(hook, ctx);
    }
  }
}

async function callHook<Ctx extends { event: string }, R>(
  hook: (ctx: Ctx) => R | Promise<R>,
  ctx: Ctx,
): Promise<R | null> {
  const start = performance.now();
  try {
    return await hook(ctx);
  } catch (err) {
    log.error('[realtime] hook failed', { hookEvent: ctx.event }, err);
    return null;
  } finally {
    const elapsedMs = performance.now() - start;
    if (elapsedMs > SLOW_HOOK_WARN_THRESHOLD_MS) {
      log.warn('[realtime] slow hook', { hookEvent: ctx.event, elapsedMs });
    }
  }
}

File: python/src/cosmo_ai/_internal/hooks.py
"""Hooks for the realtime SDK: in-process lifecycle callbacks fired by the
session at four seams (SessionStart, PreToolUse, PostToolUse, SessionEnd).

Every seam hangs off something this process does, which is what makes it a
seam rather than a notification: SessionStart rewrites the config before it
is sent, PreToolUse gates a local handler, PostToolUse reads that handler's
outcome, and SessionEnd fires even when the socket dropped and no frame ever
arrived. Anything the SERVER did arrives as an event on the session's event
stream — ``UserSpeechTimeoutEvent`` for a fired silence hook — never as a
hook.

Declare a hook with the seam's decorator — the decorated name becomes a
:class:`Hook` — and attach with ``hooks=[...]`` on
:meth:`RealtimeClient.agent`; list order is fold order::

    from cosmo_ai import hooks

    @hooks.session_start
    def add_context(ctx) -> SessionStartResult:
        return SessionStartResult(additional_context="Be concise.")

    @hooks.pre_tool_use(matcher="delete_*")
    def block_deletes(ctx) -> PreToolUseResult:
        return PreToolUseResult(permission="deny", reason="disabled")

    agent = client.agent(..., hooks=[add_context, block_deletes])

Observer-grade by default; the two client-controlled seams honor overrides —
``SessionStart`` may inject ``additional_context`` into the instructions and
``PreToolUse`` may deny or rewrite a local client-tool call. A throwing hook is
isolated and never breaks the session. A malformed matcher raises at
decoration, not at session start.
"""

from __future__ import annotations

import time
from collections.abc import Sequence
from dataclasses import dataclass
from enum import Enum
from fnmatch import fnmatchcase
from inspect import isawaitable
from types import MappingProxyType
from typing import TYPE_CHECKING, Any, Awaitable, Callable, Literal, Union, overload

import structlog

from cosmo_ai._internal.logging import get_logger
from cosmo_ai._internal.protocol import ServerHook

if TYPE_CHECKING:
    pass

logger: structlog.stdlib.BoundLogger = get_logger(__name__)


class DisconnectReason(str, Enum):
    """Why the session reached ``DISCONNECTED``. Defined here (not in the
    session engine) because the SessionEnd hook context carries it; the
    ``session`` facade re-exports it."""

    CLIENT_ENDED = "client_ended"
    CLIENT_CLOSED = "client_closed"
    HANDSHAKE_FAILED = "handshake_failed"
    SERVER_ENDED = "server_ended"
    TRANSPORT_ERROR = "transport_error"

HookEventName = Literal["SessionStart", "PreToolUse", "PostToolUse", "SessionEnd"]

# A hook runs in-process on the session's hot path (SessionStart blocks session
# establishment; PreToolUse/PostToolUse are awaited inline in the tool-call RPC
# reply during a live voice turn — see design doc §7). Nothing bounds a hook's
# runtime, so a slow hook (e.g. a network call) stalls that path; this only
# warns, it never cancels or times out a hook.
_SLOW_HOOK_WARN_THRESHOLD_S = 0.2


# ── Tool outcome (what PostToolUse observes) ───────────────────────────


@dataclass(frozen=True)
class ToolOk:
    result: dict[str, Any] | None


@dataclass(frozen=True)
class ToolError:
    message: str


@dataclass(frozen=True)
class ToolDenied:
    reason: str


ToolOutcome = Union[ToolOk, ToolError, ToolDenied]


# ── Per-event contexts ─────────────────────────────────────────────────


@dataclass(frozen=True)
class SessionStartContext:
    # session_id does not exist until the handshake completes; read ReadyEvent
    # off the event stream for the started id.
    event: Literal["SessionStart"] = "SessionStart"


@dataclass(frozen=True)
class PreToolUseContext:
    tool_name: str
    arguments: "MappingProxyType[str, Any]"  # read-only; rewrite via PreToolUseResult
    session_id: str
    event: Literal["PreToolUse"] = "PreToolUse"


@dataclass(frozen=True)
class PostToolUseContext:
    tool_name: str
    arguments: dict[str, Any]
    outcome: ToolOutcome
    session_id: str
    event: Literal["PostToolUse"] = "PostToolUse"


@dataclass(frozen=True)
class SessionEndContext:
    reason: DisconnectReason
    detail: str | None
    session_id: str | None
    event: Literal["SessionEnd"] = "SessionEnd"


HookContext = Union[
    SessionStartContext,
    PreToolUseContext,
    PostToolUseContext,
    SessionEndContext,
]


# ── Per-event results (only override-capable events) ───────────────────


@dataclass(frozen=True)
class SessionStartResult:
    additional_context: str | None = None


@dataclass(frozen=True)
class PreToolUseResult:
    permission: Literal["allow", "deny"] | None = None
    reason: str | None = None
    updated_arguments: dict[str, Any] | None = None


@dataclass(frozen=True)
class PreToolUseOutcome:
    """Folded result of all PreToolUse hooks for one tool call."""

    denied: bool
    reason: str | None
    arguments: dict[str, Any]


HookCallback = Callable[[Any], Union[Awaitable[Any], Any]]


@dataclass(frozen=True)
class Hook:
    """One declared hook: the seam it fires on, the callback, and (for the
    tool seams) the matcher. Produced by the seam decorators; attach with
    ``hooks=[...]`` — list order is fold order."""

    event: HookEventName
    callback: HookCallback
    matcher: str | None = None


def tool_name_matches(tool_name: str, pattern: str) -> bool:
    """Normative matcher grammar, shared with the Swift SDK via
    ``hook-matcher-vectors.json``: glob-style
    ``*`` ``?`` ``[seq]`` ``[!seq]``, case-sensitive, matched against the full
    tool name, no path semantics. ``fnmatchcase`` because plain ``fnmatch`` is
    case-insensitive on Windows. A malformed pattern is rejected at
    decoration (:func:`_validate_matcher`) rather than reaching this
    function, so every pattern seen here is well-formed."""
    return fnmatchcase(tool_name, pattern)


def _validate_matcher(pattern: str) -> None:
    """Reject an unterminated ``[...]`` group at decoration time.

    ``fnmatch`` never raises on a malformed pattern — it treats a stray ``[``
    as a literal character (mirrors ``fnmatch.translate``'s own bracket scan),
    so e.g. ``matcher="[delete_*"`` would silently never match any real tool
    name instead of erroring. For a ``PreToolUse`` deny matcher that is a
    silent fail-open (the guard never fires), so this fails loud instead."""
    i, n = 0, len(pattern)
    while i < n:
        if pattern[i] == "[":
            j = i + 1
            if j < n and pattern[j] == "!":
                j += 1
            if j < n and pattern[j] == "]":
                j += 1
            while j < n and pattern[j] != "]":
                j += 1
            if j >= n:
                raise ValueError(
                    f"malformed hook matcher {pattern!r}: unterminated '[' at index {i}"
                )
            i = j + 1
        else:
            i += 1


# ── Seam decorators ────────────────────────────────────────────────────


def _seam(
    event: HookEventName, fn: HookCallback | None, matcher: str | None = None
) -> Any:
    if matcher is not None:
        _validate_matcher(matcher)
    if fn is None:
        return lambda f: Hook(event, f, matcher)
    return Hook(event, fn, matcher)


@overload
def session_start(fn: HookCallback) -> Hook: ...
@overload
def session_start(fn: None = None) -> Callable[[HookCallback], Hook]: ...
def session_start(fn: HookCallback | None = None) -> Any:
    """Declare a ``SessionStart`` hook — may return a
    :class:`SessionStartResult` to inject ``additional_context``."""
    return _seam("SessionStart", fn)


@overload
def pre_tool_use(fn: HookCallback) -> Hook: ...
@overload
def pre_tool_use(*, matcher: str | None = None) -> Callable[[HookCallback], Hook]: ...
def pre_tool_use(
    fn: HookCallback | None = None, *, matcher: str | None = None
) -> Any:
    """Declare a ``PreToolUse`` hook — may return a :class:`PreToolUseResult`
    to deny or rewrite a local client-tool call. ``matcher`` restricts it to
    matching tool names; bare form matches every tool."""
    return _seam("PreToolUse", fn, matcher)


@overload
def post_tool_use(fn: HookCallback) -> Hook: ...
@overload
def post_tool_use(*, matcher: str | None = None) -> Callable[[HookCallback], Hook]: ...
def post_tool_use(
    fn: HookCallback | None = None, *, matcher: str | None = None
) -> Any:
    """Declare a ``PostToolUse`` observer, fired with the final
    :data:`ToolOutcome` of each local client-tool call."""
    return _seam("PostToolUse", fn, matcher)


@overload
def session_end(fn: HookCallback) -> Hook: ...
@overload
def session_end(fn: None = None) -> Callable[[HookCallback], Hook]: ...
def session_end(fn: HookCallback | None = None) -> Any:
    """Declare a ``SessionEnd`` observer, fired exactly once at teardown."""
    return _seam("SessionEnd", fn)


def resolve_hooks(
    hooks: Sequence[Hook | ServerHook] | None, *, server_allowed: bool = True
) -> tuple[Hook | ServerHook, ...] | None:
    """Snapshot the agent's ``hooks`` argument. Elements are :class:`Hook`
    (client callbacks, from the seam decorators) or :data:`ServerHook` kinds
    (wire config the server executes). A catalog launch passes
    ``server_allowed=False``: its stored config governs server behavior, so
    server hooks are rejected there."""
    if hooks is None:
        return None
    for hook in hooks:
        if isinstance(hook, ServerHook):
            if not server_allowed:
                raise TypeError(
                    "a catalog agent runs its stored config verbatim — "
                    "server hooks cannot ride along"
                )
        elif not isinstance(hook, Hook):
            raise TypeError(
                "hooks elements must be Hook (declare with the seam "
                "decorators) or a server hook such as SilenceTimeout, "
                f"got {type(hook).__name__}"
            )
    return tuple(hooks)


# ── Dispatch engine ────────────────────────────────────────────────────


class HookEngine:
    """Dispatch over an ordered hook tuple. Fold semantics are pinned by
    ``hook-engine-vectors.json``, shared with the
    TS and Swift SDKs."""

    def __init__(self, hooks: Sequence[Hook]) -> None:
        self._by_event: dict[HookEventName, list[Hook]] = {
            "SessionStart": [],
            "PreToolUse": [],
            "PostToolUse": [],
            "SessionEnd": [],
        }
        for hook in hooks:
            self._by_event[hook.event].append(hook)

    async def _call(self, hook: Hook, ctx: HookContext) -> Any:
        start = time.monotonic()
        try:
            out = hook.callback(ctx)
            if isawaitable(out):
                out = await out
            return out
        except Exception:
            logger.exception(
                "realtime.hook_failed", hook_event=ctx.event, stack_info=True
            )
            return None
        finally:
            elapsed_s = time.monotonic() - start
            if elapsed_s > _SLOW_HOOK_WARN_THRESHOLD_S:
                logger.warning(
                    "realtime.hook_slow",
                    hook_event=ctx.event,
                    elapsed_ms=round(elapsed_s * 1000, 1),
                )

    async def run_session_start(self, ctx: SessionStartContext) -> str | None:
        chunks: list[str] = []
        for hook in self._by_event["SessionStart"]:
            out = await self._call(hook, ctx)
            if isinstance(out, SessionStartResult) and out.additional_context:
                chunks.append(out.additional_context)
        return "\n\n".join(chunks) if chunks else None

    async def run_pre_tool_use(
        self, *, tool_name: str, arguments: dict[str, Any], session_id: str
    ) -> PreToolUseOutcome:
        current = dict(arguments)
        for hook in self._by_event["PreToolUse"]:
            if hook.matcher is not None and not tool_name_matches(
                tool_name, hook.matcher
            ):
                continue
            ctx = PreToolUseContext(
                tool_name=tool_name,
                arguments=MappingProxyType(current),
                session_id=session_id,
            )
            out = await self._call(hook, ctx)
            if not isinstance(out, PreToolUseResult):
                continue
            if out.permission == "deny":
                logger.info(
                    "realtime.hook_denied_tool", tool=tool_name, reason=out.reason
                )
                return PreToolUseOutcome(
                    denied=True, reason=out.reason or "denied by hook", arguments=current
                )
            if out.updated_arguments is not None:
                current = dict(out.updated_arguments)
        return PreToolUseOutcome(denied=False, reason=None, arguments=current)

    async def run_post_tool_use(self, ctx: PostToolUseContext) -> None:
        for hook in self._by_event["PostToolUse"]:
            if hook.matcher is not None and not tool_name_matches(
                ctx.tool_name, hook.matcher
            ):
                continue
            await self._call(hook, ctx)

    async def run_session_end(self, ctx: SessionEndContext) -> None:
        for hook in self._by_event["SessionEnd"]:
            await self._call(hook, ctx)
